src/zjs_callbacks.c

```diff
@@ -387,64 +386,86 @@ void print_callbacks(void)
-                ret_val = jerry_call_function(cb_map[i]->js->func_list[j], cb_map[i]->js->this, args, argc);
-                if (jerry_value_has_error_flag(ret_val)) {
-                    DBG_PRINT("callback %ld returned an error for function[%i]\n", i, j);
+                for (i = 0; i < cb_map[id]->js->num_funcs; ++i) {
+                    if (sz) {
+                        ret_val = jerry_call_function(cb_map[id]->js->func_list[i], cb_map[id]->js->this, data, sz);
+                    } else {
+                        ret_val = jerry_call_function(cb_map[id]->js->js_func, cb_map[id]->js->this, NULL, 0);
+                    }
+                    if (jerry_value_has_error_flag(ret_val)) {
+                        DBG_PRINT("callback %d returned an error for function[%i]\n", id, i);
+                    }
+                }
+                if (cb_map[id]->js->post) {
+                    cb_map[id]->js->post(cb_map[id]->js->handle, &ret_val);
                 }
             }
-            if (cb_map[i]->js->post) {
-                cb_map[i]->js->post(cb_map[i]->js->handle, &ret_val);
+        } else if (cb_map[id]->type == CALLBACK_TYPE_C && cb_map[id]->c->function) {
+            if (sz) {
+                cb_map[id]->c->function(cb_map[id]->c->handle, data);
+            } else {
+                cb_map[id]->c->function(cb_map[id]->c->handle, NULL);
             }
         }
-    } else if (cb_map[i]->type == CALLBACK_TYPE_C && cb_map[i]->c->function) {
-        DBG_PRINT("calling callback id %ld\n", cb_map[i]->c->id);
-        cb_map[i]->c->function(cb_map[i]->c->handle);
     }
 }
 
 void zjs_service_callbacks(void)
 {
-    int i;
-    for (i = 0; i < cb_size; i++) {
-        if (cb_map[i] && cb_map[i]->signal) {
-            cb_map[i]->signal = 0;
-            zjs_call_callback(i);
+    if (ring_buf_initialized) {
+        uint16_t type;
+        uint8_t value;
+        uint8_t size = 0;
+        uint32_t dummy;
+        while (zjs_port_ring_buf_get(&ring_buffer,
+                                     &type,
+                                     &value,
+                                     &dummy,
+                                     &size) == -EMSGSIZE) {
+            if (size == 0) {
+                break;
+            } else {
+                // Pull from ring buffer
+                uint16_t id;
+                uint8_t value1;
+                uint8_t sz = size;
+                jerry_value_t data[sz];
+
+                int ret = zjs_port_ring_buf_get(&ring_buffer,
+                                                &id,
+                                                &value1,
+                                                (uint32_t*)data,
+                                                &sz);
+                if (ret != 0) {
+                    DBG_PRINT("(JS) error pulling from ring buffer: ret = %u\n", ret);
+                    break;
+                }
+                zjs_call_callback(id, data, sz);
```

Comment:
Oops, this is a bug. If there are no args, the while will not return -EMSGSIZE... it will work! That was a gotcha I mentioned. So you need to check for that case and still call the callback. If I'm not mistaken, you're not calling any 0-arg functions right now.

(Seems surprising guess I'll go test it when I'm done here.)

So... only call buf_get the second time *IF* there were args and it was needed. But either way, call the callback.